### ISSUE TITLE
crud-web-apps use deprecated Angular 8

### ISSUE THREAD

**Issue by DavidSpek** [kubeflow/kubeflow#5525]
/kind bug
/area jupyter

The crud-web-apps, as well as the `/components/jupyter-web-app` that doesn't use the common code are using Angular 8, which is deprecated and is not receiving security patches. The current version of Angular is 11 (12 has been released now), with 9 being the last still receiving updates until August 6th, 2021. I've started an attempt to update the crud-web-apps to Angular 9 (it is not recommended to skip multiple versions at once) which went relatively smoothly for the common code. However, when updating the Jupyter web app errors arise related https://v9.angular.io/guide/migration-injectable and https://v9.angular.io/guide/migration-undecorated-classes. These errors appear during migration of the Jupyter web app and are all related to the common frontend code.

**Comment by kimwnasptd** [comment]
Thank you for starting this effort @DavidSpek.

Having this done in parallel can help us save time for when we decide we have to move on from Angular 8, but this is not the case in the immediate future. The reasons why don't want to commit huge effort on this right now are:
1. There are other UIs in the pipeline that will come within the month like the Volumes web app and the new [Katib UI](https://github.com/kubeflow/katib/issues/1322#issuecomment-762223386), which will be using this stack and common code
2. We don't have solid tests that can test if our images can be built or that, at least, some basic functionality is being tested
3. While Angular 8 will not receive security fixes in the future this doesn't mean that right now we have a hot potato of a security issue in our hands that we need to solve immediately
4. I don't have the cycles to review this until the release of 1.3 

But I agree that this is something we will have to invest cycles in the future. So if you'd like to start this effort in parallel with 1.3 release then you are more than welcome to. But we will start thinking about merging and investing more cycles in this at least afterwards 1.3. 

Also, if you plan on continuing, please document in this issue what difficulties/challenges or changes you had to make for supporting Angular 9. This will help us better grasp the effort required as well as potential pitfalls.

**Comment by DavidSpek** [comment]
@kimwnasptd I completely agree with you, and I was aware that this effort would also influence the new UIs that are being released and should be done all at once. Also, just to be clear I'd like to note that my vision of updating Angular is not to go to Angular 9, but to go to whatever the latest stable version is at that point in time. 

The issues I have had to overcome until this point should be fairly clear from the git commits, but in the future I will add relevant information for changes that are needed. The pain point I am at now is related to linking the common code to the jupyter code during the migration, which makes me wonder if it might not be advisable to have the common code as a package on npm. 

FYI, regarding security updates, the packages as they are now have some high risk security vulnerabilities according to npm and is not using the latest Angular 8 version. I think it is wise to update to the latest Angular 8 version and running `npm audit fix` and `npm update` soon.

**Comment by DavidSpek** [comment]
@kimwnasptd I've just created a PR that updates all the CRUD web apps in this repo from Angular 8 to Angular 11. I've added the command to (almost) every commit as the description so it should be easily to follow. I think the only web app that is now excluded from this is the new Katib UI. However, I believe it will be fairly easy to update that from Angular 8 to 11 using the commands in the descriptions of my commits as a guide once https://github.com/kubeflow/kubeflow/pull/5805 has been merged. I'm also willing to put in the effort to update the new Katib UI.

**Comment by kimwnasptd** [comment]
Thank you for your efforts on this @DavidSpek!

Have you bumped into blockers or critical issues that will require us to migrate to the newest version of Angular? This will help us schedule our cycles in this effort.

The main issues around the migration that we need to take in to consideration are:
1. we don't have e2e tests that constantly ensure the core functionalities of the apps work
2. we don't have unit tests for the UI code of the apps themselfes
3. we have 5 different web apps that are using the common UI code.
    * [Jupyter web app](https://github.com/kubeflow/kubeflow/tree/master/components/crud-web-apps/jupyter)
    * [Volumes web app](https://github.com/kubeflow/kubeflow/tree/master/components/crud-web-apps/volumes)
    * [Tensorboards web app](https://github.com/kubeflow/kubeflow/tree/master/components/crud-web-apps/tensorboards)
    * [Katib web app](https://github.com/kubeflow/katib/tree/master/pkg/new-ui/v1beta1)
    * [Models web app](https://github.com/kubeflow/kfserving/tree/master/web-app)
4. migrating to a version equal or bigger than Angular 9 [we use 8.2] will require significant changes, due to the [Ivy compiler](https://angular.io/guide/ivy) that was introduced in that version

I have some work on the way for unit and e2e tests, but I believe we could also put cycles into this effort in parallel as well. I think the best way forward to achieve this will be to place the newest code into directories named `frontend-v11`, or something similar, and not replace the existing code. Then once we have the e2e tests in place, and are confident with the newest code, we can make the switch.

WDYT?

**Comment by DavidSpek** [comment]
@kimwnasptd I haven't bumped into any blockers other than a bug in the upgrade process from 8 to 9. Testing the web apps manually everything works. The only thing that is not working at the moment is the CI test for the common frontend code. However, locally this works without a problem so I am not sure why a certain part of the test is failing. 

I haven't noticed any issues when using Ivy, so I think it is alright. There will surely be things that will need optimization or a small fix here or there after the upgrade, but overall the web apps to still function as expected. As you said, the difficult thing is that there is no CI in place at the moment. However, not upgrading until the CI is all in place will, I think, cause the upgrade to be postponed indefinitely resulting in more and duplicate work. Also, with Angular 11 there are some changes in parts of the testing code so that would need to migrate as well. 

I think merging it into a directory like `frontend-v11` is a good option. However, I really want to avoid the situation where the switch from 8 to 11 ends up not happening timely and complicates things even more. So, as a possible alternative to a `frontend-v11` directory I have 2 questions. 

1. What would be the process to validate the web apps manually? Is running `npm run build:watch` for the frontend with `make run-dev` for the backend then checking in the browser console for logs and testing normal functionality enough? What more would need to be done to verify the web apps otherwise?
2. Would an alternative option be to move the current folder to `frontend-v8` and have the images build from there for the time being, but having all future development work being done on the v11 web apps once they are merged?
